internal/pkg/deploy/cloudformation/app.go

```diff
@@ -313,12 +373,77 @@ func (cf CloudFormation) RemoveJobFromApp(app *config.Application, jobName strin
 	return nil
 }
 
+// RemoveEnvFromAppOpts contains the parameters to call RemoveEnvFromApp.
+type RemoveEnvFromAppOpts struct {
+	App          *config.Application
+	EnvToDelete  *config.Environment
+	Environments []*config.Environment
+}
+
+// RemoveEnvFromApp optionally redeploys the app stack to remove the account and, if necessary, empties
+// ECR repos and a regional S3 bucket before deleting the stackset instance for that region. This method
+// cannot check that deleting the stackset or removing the app won't break copilot. Be careful.
+func (cf CloudFormation) RemoveEnvFromApp(opts *RemoveEnvFromAppOpts) error {
+	var accountHasOtherEnvs, regionHasOtherEnvs bool
+	for _, env := range opts.Environments {
+		if env.Name != opts.EnvToDelete.Name {
+			if env.AccountID == opts.EnvToDelete.AccountID {
+				accountHasOtherEnvs = true
+			}
+			if env.Region == opts.EnvToDelete.Region {
+				regionHasOtherEnvs = true
+			}
+		}
+	}
+
+	// This is a no-op if there are remaining environments in the region or account.
+	if regionHasOtherEnvs && accountHasOtherEnvs {
+		return nil
+	}
+
+	deployApp := &deploy.CreateAppInput{
+		Name:               opts.App.Name,
+		AccountID:          opts.App.AccountID,
+		DomainName:         opts.App.Domain,
+		DomainHostedZoneID: opts.App.DomainHostedZoneID,
+		Version:            opts.App.Version,
+	}
+	appConfig := stack.NewAppStackConfig(deployApp)
+
+	if !regionHasOtherEnvs {
```

Comment:
I like that these are handled in the same function! It feels weird though that all these logics non-CFN are in the `deploy/cloudformation` package. Maybe this signals that we will need a refactor soon.

Reply:
Agreed, I think that perhaps this stuff needs to get pulled up one level into the deploy package. Or moved over into clideploy. It's probably worth putting all our deletion logic into that package as well as the deployment logic, now that I think about it.